### PR TITLE
Add needed speech recognition changes for android 11

### PR DESCRIPTION
**Pull request by razzeee**
## Description

As mentioned in the forum at https://forum.kodi.tv/showthread.php?tid=366478&pid=3082674#pid3082674

## Motivation and context

@Klojum posted this in slack

## How has this been tested?

Hasn't been tested, we'll probably use the android server buildt to test it.

## What is the effect on users?

Should fix android 11 I think?

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

**Comment by Tolriq** [comment]
"RECORD_AUDIO" is a dangerous permission so is not granted automatically, if it fixes something then it must be requested like when Kodi request the permission to access files.(https://github.com/xbmc/xbmc/blob/master/tools/android/packaging/xbmc/src/Splash.java.in#L142)

**Comment by razzeee** [comment]
Your obviously right, I shouldn't just blindly implement stuff people post in the forum :dagger: 
Not sure if I can find time for the correct fix.

**Comment by matthuisman** [comment]
its not as simple as adding the below?

```
  requestPermissions(new String[]{Manifest.permission.RECORD_AUDIO},
            PERMISSION_RESULT_CODE);
```

**Comment by razzeee** [comment]
Ideally we wouldn't do it on the splash and do it atomically on the operations that need it...

**Comment by matthuisman** [comment]
gotcha, so ACTION_VOICE_RECOGNIZE we want to request permission if not yet accepted

**Comment by razzeee** [comment]
realistically we want to do a similar thing on linux, but ask https://github.com/flatpak/xdg-desktop-portal/ via instead

**Comment by matthuisman** [comment]
right, that is all above my pay grade :)

the changes below should ask for audio on splash (and require both audio and storage for kodi to start)
https://github.com/xbmc/xbmc/compare/master...matthuisman:android_audio?expand=1
which is probably not wanted.

Also, i'd have no idea of how to use libandroidjni to request permissions with the callbacks etc.

**Comment by a1rwulf** [comment]
> Ideally we wouldn't do it on the splash and do it atomically on the operations that need it...

Not sure if I would agree, I think asking for the permission in Splash would be fine too.
I know the modern approach is to do it when you need access, but if I eval my kodi usage, I'd rather grant or deny when I first start the app and don't get bothered afterwards - subjective opinion of course.

**Comment by Tolriq** [comment]
From long experience this only works for permissions that makes direct sense.
If a media center ask for record audio permission without explanation, they will refuse at 100% and bad rate saying privacy invasion bla bla bla since no one would understand why this permission is needed.

**Comment by Zuikkis** [comment]
Well. Shield@NVIDIA commented on the Kodi forum:

> If I can summarize, Kodi on SHIELD  Experienc 8.2.3 (Android 9) was using the Micorphone without asking the user for Privleges and now you would like to change that with SHIELD Experience 9.0 (Android 11).

That's a good point. Why do you want to ask for permission now when you didn't in Android 9? This is especially confusing for users; a thing that worked in Android 9 now suddenly doesn't work.

Most Google guidelines are made for phones in mind. In Kodi's case RECORD_AUDIO is not so dangerous than have random app listen to you in your pocket..

**Comment by a1rwulf** [comment]
> From long experience this only works for permissions that makes direct sense. If a media center ask for record audio permission without explanation, they will refuse at 100% and bad rate saying privacy invasion bla bla bla since no one would understand why this permission is needed.

Nobody said that we shouldn't explain why we need this permission, I just stated that I personally still like it more to be asked on app startup that's all.

>That's a good point. Why do you want to ask for permission now when you didn't in Android 9? This is especially confusing for users; a thing that worked in Android 9 now suddenly doesn't work.

Because it has been made necessary by google.
If you used an iOS/Android phone from the early beginning, you know that in the early days there wasn't even a proper permission model in place, so apps could do whatever they wanted (and this was heavily exploited by a ton of apps).
Google/Apple continuously improves how those privileges should be gotten granted by users, to make it more transparent what an app can do with it and why the app needs this - it's just evolution of technology.

**Comment by Tolriq** [comment]
> Nobody said that we shouldn't explain why we need this permission, I just stated that I personally still like it more to be asked on app startup that's all.

I was probably not clear enough, I meant the 2. 
Asking upfront never works on non direct logical permissions. Users won't really read the text, just see the record audio, refuse it as it makes no sense for a media center (Specially on Android where you can get user voice input via intents and not require this permission at all).
You have no idea how many support I get from Kodi 19 upgrade that warns the users that webserver is disabled if it had no password, they just clicked ok without reading then complains that webserver no more works :)

Then comes the goldfish effect, they click on the button that require the permission, it does not work, they complain and trigger tons of support.

**Comment by Zuikkis** [comment]
> Because it has been made necessary by google.

Not really? For current Android 11, at least if running in Nvidia Shield, it is enough to just add the RECORD_AUDIO permission to the androidmanifest.xml. 

While it would be **better** to ask user for the permission, currently that simple fix would make things work the way they worked in Android 9.

**Comment by a1rwulf** [comment]
@Tolriq yeah I misunderstood before, thx for clarifying.

@Zuikkis no, not exactly true.
Before, you didn't even need to add anything to the manifest, it just worked.
Now this different: https://developer.android.com/guide/topics/media/mediarecorder#audio-record-permission

To quote google's docs:

>RECORD_AUDIO is considered a "dangerous" permission because it may pose a risk to the user's privacy. Starting with Android 6.0 (API level 23) an app that uses a dangerous permission must ask the user for approval at run time.

So here we are.
As stated in the forums, somebody needs to do it - period.
We are all volunteers, so nobody will be forced to add this, it will be broken for the time being
until someone inside team Kodi takes the task or an outside contributor will PR a proper solution.

**Comment by Zuikkis** [comment]
@a1rwulf but Shield@NVIDIA says:

> we tested Kodi against Googles adt3 development platform with Android 11 and it had the same issue. We have also made an apk with the Manifest changes and have confirmed it fixes the issue.

Has anyone else actually tried the manifest change and it didn't work?

**Comment by a1rwulf** [comment]
> @a1rwulf but Shield@NVIDIA says:
> 
> > we tested Kodi against Googles adt3 development platform with Android 11 and it had the same issue. We have also made an apk with the Manifest changes and have confirmed it fixes the issue.
> 
> Has anyone else actually tried the manifest change and it didn't work?

I'm not sure what you want, adding it to the manifest without asking the user for permission is clearly a violation of the rules.
If you want to rant about that decision, please head over to the forum, this conversation is spamming hundreds of people, so I'll stop to reply in here.

**Comment by Tolriq** [comment]
Since I still can't post on the forums so posting here ;)

I've quickly checked and there's a ton of misunderstanding here :)
1) Kodi seems to use ACTION_RECOGNIZE_SPEECH this does not require any permission unless the app do implement a RecognizerService and it's does not seems to be the case, so if there's a permission issue this is not on Kodi side.
2) Kodi is still targeting sdk 29: So the queries filter does nothing, but and that's a big big big but : `Starting in November 2021, app updates will be required to target API level 30 or above and adjust for behavioral changes in Android 11.` so Kodi needs to be updated to target SDK 30 (And probably 31 at the new target for november 2022). That update will require the queries part in the manifest, but have a tons of other impactful things on Kodi that needs to be addressed.

So TL;DR, Kodi can't be updated on Play Store without big changes as explained a few time last year. And the root issue here seems to be completely different from what most think about.

For the record Yatse does use ACTION_RECOGNIZE_SPEECH, does not use, list or require RECORD_AUDIO, target SDK 31 and works perfectly.
